### PR TITLE
Device client ArgumentException rework

### PR DESCRIPTION
**Pull request by tmahmood-microsoft**

### CODE REVIEW

File: iothub/device/src/Authentication/DeviceAuthenticationWithSharedAccessPolicyKey.cs [hunk 1]
```diff
@@ -61,10 +61,7 @@ public string PolicyName
         /// <returns>The populated <see cref="IotHubConnectionCredentials"/> instance.</returns>
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/DeviceAuthenticationWithSharedAccessPolicyKey.cs:61]: document exceptions

File: iothub/device/src/Authentication/HsmAuthentication/HttpHsmSignatureProvider.cs [hunk 2]
```diff
@@ -27,29 +28,17 @@ internal class HttpHsmSignatureProvider : ISignatureProvider
 
         public HttpHsmSignatureProvider(string providerUri, string apiVersion)
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/HsmAuthentication/HttpHsmSignatureProvider.cs:29]: Can we make any public an internal (those not inheriting from ISignatureProvider), like this ctor?

File: iothub/device/src/Authentication/IotHubConnectionCredentials.cs [hunk 3]
```diff
@@ -313,23 +313,32 @@ private void SetAuthenticationModel()
         private void Validate()
         {
             // IoT Hub host name
-            Argument.AssertNotNullOrWhiteSpace(IotHubHostName, nameof(IotHubHostName));
+            if (string.IsNullOrWhiteSpace(IotHubHostName))
+            {
+                throw new FormatException("Missing IoT Hub host name.");
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/IotHubConnectionCredentials.cs:318]:
```suggestion
                throw new FormatException("Missing IoT hub host name.");
```

File: iothub/device/src/Authentication/Security/StringValidationHelper.cs [hunk 4]
```diff
@@ -25,7 +25,7 @@ internal static void EnsureBase64String(string value, string paramName)
         {
             if (!IsBase64StringValid(value))
             {
-                throw new ArgumentException(string.Format(CultureInfo.InvariantCulture, StringIsNotBase64, value), paramName);
+                throw new InvalidOperationException(string.Format(CultureInfo.InvariantCulture, StringIsNotBase64, value, paramName));
```
Comment (drwill-ms) [inline iothub/device/src/Authentication/Security/StringValidationHelper.cs:28]: Format exception?

File: iothub/device/src/Authentication/HsmAuthentication/GeneratedCode/HttpHsmClient.cs [hunk 5]
```diff
@@ -64,14 +65,10 @@ internal HttpHsmClient(HttpClient httpClient)
         /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
         internal async Task<SignResponse> SignAsync(string apiVersion, string name, string genid, SignRequest payload, CancellationToken cancellationToken = default)
         {
-            if (name == null)
-                throw new ArgumentNullException("name");
+            Debug.Assert(name != null, $"{nameof(name)} cannot be null. Validate parameter upstream.");
+            Debug.Assert(genid != null, $"{nameof(genid)} cannot be null. Validate parameter upstream.");
+            Debug.Assert(apiVersion != null, $"{nameof(apiVersion)} cannot be null. Validate parameter upstream.");
```
Comment (abhipsaMisra) [inline iothub/device/src/Authentication/HsmAuthentication/GeneratedCode/HttpHsmClient.cs:70]: 👍🏽

**Review by abhipsaMisra — APPROVED** [review]

**Pull request merged**